flang/lib/Lower/OpenACC.cpp

```diff
@@ -15,38 +15,57 @@
-  std::visit(
-      common::visitors{
-          [&](const Fortran::parser::OpenACCBlockConstruct &blockConstruct) {
-            TODO();
-          },
-          [&](const Fortran::parser::OpenACCCombinedConstruct
-                  &combinedConstruct) { TODO(); },
-          [&](const Fortran::parser::OpenACCLoopConstruct &loopConstruct) {
-            TODO();
-          },
-          [&](const Fortran::parser::OpenACCStandaloneConstruct
-                  &standaloneConstruct) { TODO(); },
-          [&](const Fortran::parser::OpenACCRoutineConstruct
-                  &routineConstruct) { TODO(); },
-          [&](const Fortran::parser::OpenACCCacheConstruct &cacheConstruct) {
-            TODO();
-          },
-          [&](const Fortran::parser::OpenACCWaitConstruct &waitConstruct) {
-            TODO();
-          },
-          [&](const Fortran::parser::OpenACCAtomicConstruct &atomicConstruct) {
-            TODO();
-          },
-      },
-      accConstruct.u);
+    const Fortran::parser::OpenACCLoopConstruct &loopConstruct) {
+
+  const auto &beginLoopDirective =
+      std::get<Fortran::parser::AccBeginLoopDirective>(loopConstruct.t);
+  const auto &loopDirective =
+      std::get<Fortran::parser::AccLoopDirective>(beginLoopDirective.t);
+
+  if (loopDirective.v == llvm::acc::ACCD_loop) {
+    auto &firOpBuilder = absConv.getFirOpBuilder();
+    auto currentLocation = absConv.getCurrentLocation();
+    auto insertPt = firOpBuilder.saveInsertionPoint();
+    llvm::ArrayRef<mlir::Type> argTy;
+    mlir::ValueRange range;
+    llvm::SmallVector<int32_t, 2> operandSegmentSizes(2 /*Size=*/,
+                                                      0 /*Value=*/);
+    auto loopOp =
+        firOpBuilder.create<mlir::acc::LoopOp>(currentLocation, argTy, range);
+    loopOp.setAttr(mlir::acc::LoopOp::getOperandSegmentSizeAttr(),
+                   firOpBuilder.getI32VectorAttr(operandSegmentSizes));
+    loopOp.getRegion().push_back(new Block{});
+    auto &block = loopOp.getRegion().back();
+    firOpBuilder.setInsertionPointToStart(&block);
+    // ensure the block is well-formed.
+    firOpBuilder.create<mlir::acc::YieldOp>(currentLocation);
+
+    // Add attribute extracted from clauses.
+    const auto &accClauseList =
+        std::get<Fortran::parser::AccClauseList>(beginLoopDirective.t);
+    for (const auto &clause : accClauseList.v) {
+      if (const auto *collapseClause{
+              std::get_if<Fortran::parser::AccClause::Collapse>(&clause.u)}) {
+
+        const auto *expr = Fortran::semantics::GetExpr(collapseClause->v);
+        const auto collapseValue = Fortran::evaluate::ToInt64(*expr);
+        if (collapseValue.has_value()) {
+          loopOp.setAttr(mlir::acc::LoopOp::getCollapseAttrName(),
+                         firOpBuilder.getI64IntegerAttr(collapseValue.value()));
+        }
+      }
+    }
+
+    // Place the insertion point to the start of the first block.
+    firOpBuilder.setInsertionPointToStart(&block);
```

Comment:
This seems odd/redundant, insertionpoint is already set at line 48. Or I'm missing something ?

Reply:
After the creation/insertion of `acc.yield` the insertion point is set after it. We want the loop to be generated before `acc.yield` so the insertionPoint is set back to the beginning of the block.

Reply:
> After the creation/insertion of `acc.yield` the insertion point is set after it. We want the loop to be generated before `acc.yield` so the insertionPoint is set back to the beginning of the block.

Thanks for clarifying :)  just noticed the yield op insertion after setting first insertionpoint.